metaflow/plugins/pypi/conda_environment.py

```diff
@@ -0,0 +1,404 @@
+        for solver in ["conda", "pypi"]:
+            with ThreadPoolExecutor() as executor:
+                results = list(
+                    executor.map(lambda x: solve(*x, solver), environments(solver))
+                )
+            _ = list(map(lambda x: self.solvers[solver].download(*x), results))
+            with ThreadPoolExecutor() as executor:
+                _ = list(
+                    executor.map(lambda x: self.solvers[solver].create(*x), results)
+                )
+            if self.datastore_type not in ["local"]:
+                # Cache packages only when a remote datastore is in play.
+                # Avoiding circular imports.
+                from metaflow.plugins import DATASTORES
+
+                storage = [d for d in DATASTORES if d.TYPE == self.datastore_type][0](
+                    _datastore_packageroot(self.datastore_type)
+                )
+                cache(storage, results, solver)
+        echo("Virtual environment(s) bootstrapped!")
+
+    def executable(self, step_name, default=None):
+        step = next(step for step in self.flow if step.name == step_name)
+        id_ = self.get_environment(step).get("id_")
+        if id_:
+            # bootstrap.py is responsible for ensuring the validity of this executable.
+            # -s is important! Can otherwise leak packages to other environments.
+            return os.path.join(id_, "bin/python -s")
+        else:
+            # for @conda/@pypi(disabled=True).
+            return super().executable(step_name, default)
+
+    def interpreter(self, step_name):
+        step = next(step for step in self.flow if step.name == step_name)
+        id_ = self.get_environment(step)["id_"]
+        # User workloads are executed through the conda environment's interpreter.
+        return self.solvers["conda"].interpreter(id_)
+
+    @functools.lru_cache(maxsize=None)
+    def get_environment(self, step):
+        environment = {}
+        for decorator in step.decorators:
+            # @conda decorator is guaranteed to exist thanks to self.decospecs
+            if decorator.name in ["conda", "pypi"]:
+                # handle @conda/@pypi(disabled=True)
+                disabled = decorator.attributes["disabled"]
+                if not disabled or str(disabled).lower() == "false":
+                    environment[decorator.name] = {
+                        k: decorator.attributes[k]
+                        for k in decorator.attributes
+                        if k != "disabled"
+                    }
+                else:
+                    return {}
+        # TODO: Support dependencies for `--metadata`.
+        # TODO: Introduce support for `--telemetry` as a follow up.
+        # Certain packages are required for metaflow runtime to function correctly.
+        # Ensure these packages are available both in Conda channels and PyPI
+        # repostories.
+        pinned_packages = {"requests": ">=2.21.0"}
```

Comment:
yes - we will call it out as a breaking change in the release notes and provide a mechanism to inject packages as a follow up.

Reply:
but why? Why break it when there is no good reason for it? What is the motivation behind this? I am the first to say "breaking changes are fine" if there is an actual reason for it. It doesn't seem like there is a good reason here since the result is pretty much the same as calling the function.